core/src/main/java/io/grpc/MessageEncoding.java

```diff
@@ -108,36 +118,93 @@ public InputStream decompress(InputStream is) throws IOException {
 
   /**
    * Registers a decompressor for both decompression and message encoding negotiation.
+   *
+   * @param d The decompressor to register
+   * @param advertised If true, the message encoding will be listed in the Accept-Encoding header.
    * @throws IllegalArgumentException if another compressor by the same name is already registered.
    */
-  public static final void registerDecompressor(Decompressor d) {
-    Decompressor previousDecompressor = decompressors.putIfAbsent(d.getMessageEncoding(), d);
-    checkArgument(previousDecompressor == null,
-        "A decompressor was already registered: %s", previousDecompressor);
+  @ExperimentalApi
+  public static final void registerDecompressor(Decompressor d, boolean advertised) {
+    DecompressorInfo previousInfo =
+        decompressors.putIfAbsent(d.getMessageEncoding(), new DecompressorInfo(d, advertised));
+    if (previousInfo != null) {
+      throw new IllegalArgumentException(
+          "A decompressor was already registered: " + previousInfo.decompressor);
+    }
   }
 
   /**
    * Provides a list of all message encodings that have decompressors available.
    */
-  public static Collection<String> getKnownMessageEncodings() {
+  @ExperimentalApi
+  public static Set<String> getKnownMessageEncodings() {
```

Comment:
Splitting out is a good idea.  How about the following:

- Codec is a top level, implements Co and Dec
- None is moved to Codec
- DecompressorRegistry is split out, with associated methods


The reason I would like to keep registry Dec specific is because there could be lot's of compressors by the same name (Gzip 9 and Gzip 3 would both make "gzip" message encodings).  Mixing them together and then needing to split them apart later would be hard.

Also, up until now I have not been entirely sure that every message encoding is also a compression method.  Any other sort of data transformation, perhaps something that interacts with the message header bits, would need to go in here.  This is very unlikely in the near future, but I was trying to be a couple years out safe.